---
title: "Dynamical-Systems View of Cell Biology"
author: "Davila-Velderrain J"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  prettydoc::html_pretty:
    theme: tactile
    highlight: github
---
This is a vignette prepared to accompany the text in this Book ... It is based on our works: [Davila-Velderrain et a. 2015a](http://journal.frontiersin.org/article/10.3389/fgene.2015.00160/full),
[2015b](http://bmcsystbiol.biomedcentral.com/articles/10.1186/s12918-015-0166-y),
[2015c](http://arxiv.org/abs/1510.04230), and
[2015d](http://link.springer.com/protocol/10.1007/978-1-4939-2444-8_23). 

## Introduction and Definitions

In systems biology, a cell is usually modeled as a dynamical system. It is assumed that the state of the cell (i.e., what is inside of it) at a certain time can be described by a set of time-dependent variables. This state is commonly represented by the amount of the different proteins within the cell or, and the latter, for practical reasons, is usually approximated by the level of gene expression (e.g., mRNA concentration). One may question the sufficiency of such representations (i.e., expression profiles), given that we now have the technical means to systematically measure a lot of other things inside the cells. I will stick here with the gene expression profile approximation to the cell state, nonetheless.

### The cell state

Thus, at any time $t$, the cell is represented by a vector $\mathbf{x}(t)=[x_1(t),...,x_n(t)]$ composed by $n$ genes (or molecules in general). Each gene $i$ has a state of activity $x_i$ associated with it, and the set of states of all the genes, the vector $\mathbf{x}$, represents the state of activity of the cell (*cell state*).   

### The state space
It is useful to imagine an abstract space termed the *state space*, where all the theoretically possible states are geometrically arranged: `each point in this abstract space represents one particular expression profile`.

### Dynamics

An important assumption implicit in all this view is that the cell state at a certain time and the cell state at a later time are connected by a state trajectory in a causal way. Thus, we can use a dynamical equation to represent a mapping from an initial state to the current cell state:

$$ \mathbf{x}(t + \delta t) = \mathbf{F}(\mathbf{x}(t) \mathbf{u}, \delta t) $$

* $\mathbf{F}$ is the transition map dynamically connecting the states
* $\mathbf{x}(t)$ is the state at time 
* $\mathbf{u}$ is a list of parameters

The transition map $\mathbf{F}$ models the regulatory circuitry of the cell -- i.e., the architecture of the `gene regulatory network (GRN)`, which specifies both the topology of the network and the nature and form of the corresponding gene regulations.

### Attractors

A set of cell states satisfy the constraints imposed by the GRN; that is, each of these cell states is connected to itself by themap

The steady states $\mathbf{x}^*$ that satisfy the the regulatory constraints of the GRN (i.e. $\mathbf{x}^* = \mathbf{F}(\mathbf{x}^* \mathbf{u})$)

When these steady states (xâˆ—) are also resilient to perturbations, that is, if they return back to the steady state after being kicked away by state variations either of intrinsic or external origin, we refer to them as attractors.

An attractor can also be composed of several quasi-stationary states repeated in a cyclic manner.

The theory posits that attractor states correspond to the observ- able robust cell phenotypes, cell types, or cellular processes; and that these emerge as a natural consequence of the dynamical con- straints imposed by the underlying GRN.


# Deterministic Dynamical Systems

## Boolean Gene Regulatory Networks

In the Boolean mathematical formalism the time increments $\delta t$ and the state variables $x_i(t)$ of the transition map $\mathbf{F}$ are discrete -- Specifically, the map takes the form

$$x_i(t+1) = F_i(x_1(t),x_2(t),...,x_k(t)),$$

* $F_i$ are logical propositions (or truth tables) expressing the relationship between the genes that share regulatory interactions with the gene $i$
* $x_i(t)$ only take the discrete values $1$ or $0$ indicating whether the gene $i$ is expressed or not


### Dynamical analysis

```{r}
library(BoolNet)
```

```{r, echo=FALSE}
setwd("/home/desoj/Documents/SIDE_B/UNAM/Libro_Springer/Vignettes")
```

```{r, echo=FALSE}
####################################################################
# FUNCTIONS
####################################################################
AttrTable <- function(Net) {
  #Extracts the attractors in a table form from the attractor object created by BoolNet function "getAttractors"
  attrs <- getAttractors(Net)  
  Attrs <- print(plotAttractors(attrs))
  Attrs <- Attrs[[1]]
  return(Attrs)
}
####################################################################
 Get.Attractors.Landscape <- function(Network) {
   attrs <- getAttractors(Network)
   TransTable <- getTransitionTable(attrs)
   
   Inits <- TransTable[[1]]
   for(i in 2:length(Network$genes)) Inits <- cbind(Inits, TransTable[[i]])
   
   AttractsClusterVector <- TransTable$attractorAssignment
   
   return(list(Inits, AttractsClusterVector))
 }
####################################################################
Get.Attractors.Landscape.Mapping <- function(Network) {
   attrs <- getAttractors(Network)
   TransTable <- getTransitionTable(attrs)
   
   Inits <- TransTable[[1]]
   for(i in 2:length(Network$genes)) Inits <- cbind(Inits, TransTable[[i]])
   
   AttractsClusterVector <- TransTable$attractorAssignment
   
   Inits
   
   InitsVec <- apply(Inits, 1, function(i) paste(i, collapse =""))
   
   return(cbind(State=InitsVec, Attractor=AttractsClusterVector))
}
####################################################################
Get.Complete.OneStep.Mapping <- function(Network) {
  Net <- Network
  attrs <- getAttractors(Net)  
  TransTable <- getTransitionTable(attrs)
  Ngenes <- length(Net$genes)
  Inits <- sapply(1:Ngenes, function(i) TransTable[[i]])
  Nstates <- sapply((Ngenes+1):(Ngenes+Ngenes), function(i) TransTable[[i]])
  Mapping <- cbind(StateT0=apply(Inits, 1, function(i) paste(i, collapse="")), StateT1=apply(Nstates, 1, function(i) paste(i, collapse="")))
  return(Mapping)
}
```


In order to analyse the Boolean model we will first read a network into `R` using the `BoolNet` package

```{r}
library(BoolNet)
Net <- loadNetwork("FOS_GRNBool.txt")
Net
```

One way to analyse a Boolean GRN model is by exhaustive computational characterization of the state space in terms of attained attractors and their basins of attractions. We can find all the attractors recvered by the GRN using a single function from `BoolNet`

```{r}
attrs <- getAttractors(Net)  
attrs
```

for easier intepretation we can plot them in matrix form

```{r}
plotAttractors(attrs)
```

we can also get the complete one-step mapping, taking all the possible states one by one as initial states

```{r}
Mapping <- Get.Complete.OneStep.Mapping(Net)
head(Mapping)
```





















# Discrete Stochastic Dynamics

### Including Stochasticity

Following \cite{alvarez2008floral, Azpeitia2014FlowerDev, davila2015modeling}, a Boolean GRN is extended into a discrete stochastic model by means of  the so--called stochasticity in nodes (SIN) model. In this model, a constant probability of error $\xi$ is introduced for the deterministic Boolean functions as follows:

$$P_{x_i(t+1)}[F_i(\mathbf{x}_{reg_i}(t))] = 1- \xi, \\$$ 
$$P_{x_i(t+1)}[1 - F_i(\mathbf{x}_{reg_i}(t))] = \xi.$$

It is assumed that the probability that the value of the random variable $x_i(t+1)$ (a gene) is determined or not by its associated logical function $F_i(\mathbf{x}_{reg_i}(t))$ is $1- \xi$ or $\xi$, respectively. The probability $\xi$ is a scalar constant parameter acting independently per gene. The vector $\mathbf{x}_{reg_i}$ represents the regulators of gene $i$.

### Inter-Attractor Transition Probability Estimation

An attractor transition probability matrix $\Pi$ with components: 

$$\pi_{ij} = P(A_{t+1}=j|A_t=i),$$

representing the probability that an attractor $j$ is reached from an attractor $i$ is estimated by either of two simulation-based algorithms proposed herein (see results). 

The elements $\pi_{ij}$ of the matrix $\Pi$ are obtained as maximum likelihood estimates based on the empirical transition probability resulting from the simulations from either algorithm 1 or 2. 

### Temporal Evolution of States/Attractors Probability

A sequence of random variables $\{C_t : t \in \mathbb{N}\}$ is considered a Markov chain (MC). $C_T$ takes as values the different attractors, the elements $\pi_{i,j}$ representing inter-attactor transition probabilities, and the matrix $\Pi$ the (one-step) transition probability matrix. As the probabilities do not depend on time, the MC is homogeneous.

The occupation probability distribution $P(C_t = j)$ -- i.e., the probability that the chain is in state (attractor or state) $j$ at a given time $t$ -- is denoted by the row vector $\mathbf{u}(t)$. The probabilities temporally evolve according to the dynamic equation

$$\mathbf{u}(t+1) = \mathbf{u}(t) \mathbf{\Pi}.$$

Taking $\mathbf{u}(0)$ as the initial distribution of the MC, the equation reads $ \mathbf{u}(1) = \mathbf{u}(0) \mathbf{\Pi}.$ By linking the occupation probabilities iteratively we get $ \mathbf{u}(t) = \mathbf{u}(0) \mathbf{\Pi}^t$: the occupation probability distribution at time $t$ can be obtained directly by matrix exponentiation.   

### Temporal-order of Attractor Attainment

Having obtained the temporal evolution of the occupation probability distribution $\mathbf{u}(t)$ given an initial distribution $\mathbf{u}(0)$ by numerically solving Equation (8); following \cite{alvarez2008floral}, it is assumed that the most likely time for an attractor to be reached is when the probability of reaching that particular attractor is maximal. Therefore, the temporal sequence in which attractors are attained is obtained by determining the sequence in which their maximum probabilities are reached using $\mathbf{u}(t)$.  

### Probabilistic Landscape

A stationary probability distribution of a MC is a distribution $\mathbf{u}_{ss}$ which satisfies the steady state equation $\mathbf{u}_{ss} = \mathbf{u}_{ss} \mathbf{\Pi}$. The stationary probability distribution, if exists, is calculated either by solving the equation $\mathbf{u}_{ss}(\mathbf{I}-\mathbf{\Pi})=0$, where $\mathbf{I}$ is the $n \, \times \, n$ identity matrix \cite{wilkinson2011stochastic}; or by numerically solving Equation (8), as $\mathbf{u}_{ss}$ corresponds to the {\em long-run distribution} of the MC: $\mathbf{u}_{ss} = \lim_{t \to \infty} \mathbf{u}(t)$ \cite{bolstad2011understanding}. A probabilistic landscape $U$ -- also called a quasi-potential -- can be obtaining by mapping the distribution $\mathbf{u}_{ss}$ using $-ln(\mathbf{u}_{ss})$. Such landscape reflects the probability of states and it provides a global characterization and a stability measure of the GRN system \cite{wang2015landscape}.

